### PR TITLE
#51/Add documentation for verification api

### PR DESCRIPTION
**Pull request by SebastianRehfeldt**
# Pull Request

### Description

This PR adds a document with general documentation on the DiGA api. It also reformats the `ENDPOINT_STATUS` file using tables for better readability.

Goal of this documentation is to help other DiGA manufacturers to understand what steps are required to verify prescription codes and how to send invoices via api requests. It also adds an FAQ section with additional questions.
Given that we at [xbird](https://xbird.io/) are at the beginning of prescription code integration, not everything in the documentation might be correct or very detailed (especially billing and encryption). Some questions in the FAQ section are also unanswered.

### TODOs

The documentation could be split up into multiple files and combined with the FAQ from the readme. Also links from the Readme to this documentation could be added before merging. Regarding our initial plan of transferring the document to the wiki, I thought I keep it in one file to reduce the work on your side.

One advantage of keeping the `docs` folder in the repository which we didnt consider in the issue is that it will be easier to apply/review changes there. Therefore, might be worth to reconsider and keep the files in the repo.

### Closes

https://github.com/alex-therapeutics/diga-api-client/issues/51

### CODE REVIEW

File: docs/diga_api.md
Comment (SebastianRehfeldt) [inline docs/diga_api.md:183]: I am not 100% about this TBH. We dont have the client up and running yet and we dont have the certificates for encryption/decryption. Therefore, it's hard to give real examples. Would be great if this could be verified or correct. Same for the response data.

File: docs/diga_api.md
Comment (SebastianRehfeldt) [inline docs/diga_api.md:198]: Or is there a way to make test requests without IK number and certificate?

File: docs/diga_api.md
Comment (SebastianRehfeldt) [inline docs/diga_api.md:220]: I was a bit shorter on this section as we will first focus on the verification. Also the xml files are more complex I think. Would be great to add some example request/response data here as well.

File: docs/diga_api.md
Comment (SebastianRehfeldt) [inline docs/diga_api.md:238]: We might add some more details when we are have successfully requested the certificates

File: docs/diga_api.md
Comment (SebastianRehfeldt) [inline docs/diga_api.md:253]: Is it really that bad?

File: docs/diga_api.md
Comment (SebastianRehfeldt) [inline docs/diga_api.md:283]: According to [this document](https://www.gkv-datenaustausch.de/media/dokumente/leistungserbringer_1/digitale_gesundheitsanwendungen/technische_anlagen_aktuell_7/DiGA_Anlage_1_Technische_Anlage_zur_RL_V1.2_20210525.pdf) on page 9, the characters can be of `[A-Z and 2-7]`

Therefore, I assume codes should be send in capitalized form.

**Comment by fongie** [comment]
Thanks! I am pretty busy right now, so just so you know it might take a few days before I get around to reviewing this.

File: docs/diga_api.md
Comment (SebastianRehfeldt) [inline docs/diga_api.md:134]: That might not necessarily be true. I found that this can be also requested from some accounting tools where a `Betriebsnummer` could be enough. Will also add some more details on how long the request for IK number takes. Initial reply says 7-10 days.

File: docs/diga_api.md
Comment (fongie) [inline docs/diga_api.md:3]: After going through the document I think I would like it to be clarified at the top what this document is for and what it is not for. Something like:

```suggestion
This documents contains information for DiGA manufacturers on how the DiGA API works regarding prescription code validation and billing. It is targeted at manufacturers who wish to develop their own integration with the DiGA API or are curious to know how it works. If you are looking for an out-of-the-box solution that already works, you can use the [diga api client](https://github.com/alex-therapeutics/diga-api-client) provided in this repository as a java library, or you can use the [diga api service](https://github.com/gtuk/diga-api-service) (work in progress), which is an api wrapper around this library, as a microservice in your stack. Information on how to use these solutions is provided in their readmes and code documentation, and not in this document.
```

just to make sure you dont end up reading this thinking you will find out how to use the library (you should read the README for that) or just looking to use something that works already

**Review by fongie — CHANGES_REQUESTED** [review]
thanks for your work on documenting this! this is information I would have loved to have from the start when I developed this integration.

Sorry that it took some time to get to reviewing this. I have requested some changes.

I am still not sure about whether to pull this into a docs folder or in the wiki. I am also a little hesitant to merge this pull request until the information it contains covers all the same use-cases that the library does. For example, the library currently fully supports billing and if you look in the source code you can find out which IK fields to put where etc., but this documentation lacks this information. I don't want someone to be confused to think that the library doesn't support billing because this documentation doesn't cover that part. The best thing would of course be for me to write that part, but unfortunately I don't have time to do that and to be honest I was kind of thinking that the code and the code documentation would be enough for now on my part in documenting this, hehe.

In short, I think that this will be a useful addition to the documentation, but you should first get your integration working and complete the incomplete sections before I will look at accepting the PR. Let me know if you have comments on this :)

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md:3]: yes, makes sense! :)

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md:51]: Thanks for pointing this out. Totally forgot about those 3 numbers.

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md:76]: Tried to translate this to English:

>  Der Fehler wird ausgegeben wenn bspw. die Zuordnung des Freischaltcodes / Rezeptcode zur DiGA nicht stimmt.

E.g. trying to use a valid code for DiGA A for DiGA B. That is the code is valid but linked to another DiGA (app). But also fine to apply your suggestion

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md:84]: Uhh, I wanted to actually do that. Will check if it's written down somewhere else but here should be the place for it

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md:91]: yes, I can have a look but also not sure when I will have time. Could it make sense to add follow-up issues for the documentation and add details piece by piece? (At least if I cannot find an answer quickly.) Could also add TODOs to the documentation to point out that some pieces are missing

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md:111]: Ahh true, had the diga api client code in mind when writing this. Will remove the library specific details and add the info about `Versandart`

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md:122]: Yes, it is self-explanatory. But I tried to keep this document on a generic level without code specifics. There is actually a section later which lists the values with explanation. I will probably reference it here.

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md:134]: Sure! :)
I was only hoping that we can reduce the waiting time by requesting the IK number and the certificates from ITSG in parallel. Maybe for test requests it would have worked with a fake IK number (of the DiGA manufacturer)? So far we are blocked by the prerequisites which also makes writing the documentation harder.

Regarding the comment on `Betriebsnummer`, it seems like there are two cases for requesting the certificates:
- as `Sonstige Leistungserbringer` with IK number - that should be our case
- as `Arbeitgeber` with `Betriebsnummer`

So I doubt that we can speed this up...

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md:198]: This is what I expected :-/ It makes it quite tough to get started with the DiGA api and this library.

Therefore, it could make sense to outline this in the beginning that these two things are blockers and should be requested early. Could even make sense in the README under prerequisites as first point?

```
Sending requests to the DiGA api requires companies to request an IK number and certificates for encryption/decryption from ITSG beforehand. These things cannot be requested in parallel and it take up to three weeks to get them. Therefore, it makes sense to request them as early as possible. Please see the {wiki | docs folder} for more information.
```

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md:220]: Yes, I also found the one from the test but wanted to use some xml files from the "real" api. But I am still blocked by the IK number and certificates and therefore, cannot send test requests.

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md:223]: I think the goal should definitely be sending invoices via api and then save invoices to the accounting software. The section was more targeted to companies which just get started with the diga api and therefore, might want to start with validation only in a first iteration. I probably make this a bit clearer in the documentation and add the point that each code requires a separate invoice.

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md:224]: Meant more that the library allows users to take manual action like sending emails. Should be clear that the library doesnt send those. But probably obsolete given your last point that no insurance uses emails.

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md:253]: We found this information [here](https://www.gkv-datenaustausch.de/media/dokumente/leistungserbringer_1/digitale_gesundheitsanwendungen/technische_anlagen_aktuell_7/DiGA_Anlage_1_Technische_Anlage_zur_RL_V1.1_20210225.pdf) in section 5.6.5.

Based on all the other documentation we understood that the mean response is 10s which was a bit of shocker. Glad to here it's not the case :)

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md:269]: Sure will add the link. It is just not mentioned as it wasn't there when I created this PR :)

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md:283]: Will do as soon as we have all the prerequisites :)
I guess for the mapping file it could cause troubles when the code is entered in lower-case. But that could probably be fixed in this library. Generally, it could make sense to capitalize the code in this library?

**Comment by SebastianRehfeldt** [comment]
> Sorry that it took some time to get to reviewing this. I have requested some changes.

No problem :) Will try to address those as soon as I have time.

> I am still not sure about whether to pull this into a docs folder or in the wiki.

Yeah it would be so great if the wiki would allow for PRs... The two additional things which came to my mind when writing the documentation:
- it will be easier to create PRs with links between the different documents if they will not be copied to the wiki after merging
- reviewing changes to the documentation will be more convenient in docs folder.

But your call and I generally agree that the wiki is the place for general documentation :)

> I am also a little hesitant to merge this pull request until the information it contains covers all the same use-cases that the library does. [...] I don't want someone to be confused to think that the library doesn't support billing because this documentation doesn't cover that part. 

Sure, the problem is that I cannot really test out the library yet and therefore struggle with describing all the use cases. Might also take a while until I can come back with the details. Another option would be to add follow-up issues and clarify in the documentation that the library is actually ahead of the documentation and has full billing support. But keeping this PR on hold would also work for me.

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md:205]: Updated the `Get Started` section in the readme.

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md]: Removed this section now

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md]: Removed this question and extended the section on the code structure. given that the code regex is described in the document, I really assume the code must be capitalized.

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md]: done

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md]: done

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md]: added the link

**Comment by SebastianRehfeldt** [comment]
Hi @fongie, 
I had some time today to address comments on the PR but there are still some things left:
  - checkout which ik field from mapping file is used where
  - add details on billing request
  - add details on ik number request (we still wait for that)
  - add details on requesting certificates from ITSG + encryption (we are still blocked)

File: ENDPOINT_STATUS.md
Comment (bokchan) [reply ENDPOINT_STATUS.md:6]: Could it make sense to extract the patch for ENDPOINT_STATUS.md into another MR? Seems like it can be merged in to main independently from changes in the two other files.

Comment (fongie) [reply ENDPOINT_STATUS.md:6]: yes @bokchan that would be nice :)

Comment (bokchan) [reply ENDPOINT_STATUS.md:6]: Ok. You wanna have the honor @SebastianRehfeldt ? :)

File: docs/diga_api.md
Comment (bokchan) [inline docs/diga_api.md:253]:
```suggestion

A [short guide](https://www.itsg.de/wp-content/uploads/2020/11/Trust-Center-howto_p10_openssl_rsa4096.pdf) (in German) for how to generate the public/private key pair with OpenSSL is available on the ITSG website under the [FAQ: _Was versteht man unter einem "Hashcode oder Fingerprint des öffentlichen Schlüssels"?_](https://www.itsg.de/produkte/trust-center/fragen-und-antworten-faq/). It also includes the command for how to export the digest/fingerprint of the public key which must be included in the application to ITSG.

Once the certificates are issued, checkout the [prerequisites from the Readme](https://github.com/alex-therapeutics/diga-api-client#prerequisites) to be able to use them with the api client.
```

File: ENDPOINT_STATUS.md
Comment (SebastianRehfeldt) [reply ENDPOINT_STATUS.md:6]: Sure thing: https://github.com/alex-therapeutics/diga-api-client/pull/59

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md:96]: Searched the codebase and both `IK_des_Rechnungsempfaengers` and `IK_Abrechnungsstelle` are only to be found in `EDFC0-basis_2.0.0.xsd` and `DigaHealthInsuranceDirectoryFromXmlTest.java`. Therefore, assuming we always have to use `Kostentraegerkennung` which is also what I found in the documentation and which is mentioned as such in the PR

Comment (SebastianRehfeldt) [reply docs/diga_api.md:96]: Oh wait! The `IK_Abrechnungsstelle` is the `clearingCenterIKNumber` ik number in the codebase and used for requests. `IK_des_Rechnungsempfaengers` is named `buyerCompanyCreditorIk` and used in billing request. The `clearingCenterIKNumber` is also used in the validation endpoint.

Also the `clearingCenterIKNumber` is the ik number of the api provider, e.g. bitmarck. Will extend the documentation next week. Last todo besides this is the billing request which we will add when we have the certificates from itsg

File: docs/diga_api.md
Comment (bokchan) [inline docs/diga_api.md:151]: I see you are you using your vacation for sensible things @SebastianRehfeldt ;)

Comment (SebastianRehfeldt) [reply docs/diga_api.md:151]: Of course! But you can see on the Todos that I already called it :) 
Gonna finish the PR when I come back. At least clarified some things around the IK numbers which were initially wrong in this file

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md:275]: I got the diga-api-client up and running locally and extracted the post data during debugging. However, I failed to send the requests and got the following traceback. Any idea how to solve?

```
java.net.ProtocolException: expected the Content-Type to have a boundary parameter
        at okhttp3.MultipartReader.<init>(MultipartReader.kt:87)
        at com.alextherapeutics.diga.implementation.DigaOkHttpClient.parseResponse(DigaOkHttpClient.java:79)
        at com.alextherapeutics.diga.implementation.DigaOkHttpClient.post(DigaOkHttpClient.java:70)
        at com.alextherapeutics.diga.DigaApiClient.performCodeValidation(DigaApiClient.java:226)
        at com.alextherapeutics.diga.DigaApiClient.sendTestCodeValidationRequest(DigaApiClient.java:153)
        at com.alextherapeutics.diga.Main.main(Main.java:57)
11:18:43.767 [main] ERROR com.alextherapeutics.diga.DigaApiClient - Failed to validate DiGA code 77AAAAAAAAAAAAAX
com.alextherapeutics.diga.DigaHttpClientException: Diga Http Client error
        at com.alextherapeutics.diga.implementation.DigaOkHttpClient.post(DigaOkHttpClient.java:73)
        at com.alextherapeutics.diga.DigaApiClient.performCodeValidation(DigaApiClient.java:226)
        at com.alextherapeutics.diga.DigaApiClient.sendTestCodeValidationRequest(DigaApiClient.java:153)
        at com.alextherapeutics.diga.Main.main(Main.java:57)
Caused by: java.net.ProtocolException: expected the Content-Type to have a boundary parameter
        at okhttp3.MultipartReader.<init>(MultipartReader.kt:87)
        at com.alextherapeutics.diga.implementation.DigaOkHttpClient.parseResponse(DigaOkHttpClient.java:79)
        at com.alextherapeutics.diga.implementation.DigaOkHttpClient.post(DigaOkHttpClient.java:70)
        ... 3 common frames omitted
```

This also happened when I started the `diga-api-service` locally. Using the service with the provided dockerfile I was able to make requests. This is where I got the response data from.

File: docs/diga_api.md
Comment (SebastianRehfeldt) [reply docs/diga_api.md:100]: Resolving this as I added some details and provided request/response examples

**Comment by SebastianRehfeldt** [comment]
@fongie it took quite long but I am done now with the requested changes :)

Thanks again for providing the client. Was able now to make some successful test requests!

**Comment by fongie** [comment]
Thanks @SebastianRehfeldt and good to hear you are up and running! I am on vacation and will get to this later in August. :)

**Comment by mkhq** [comment]
@SebastianRehfeldt @fongie Thanks for making this documentation + library, its a massive help! This is really well structured information on both the verification and billing process.

One thing that was a bit unclear to me are the options for when requests are failing, e.g. code 201, 202, connection drops, no response, ... . Can we retry both the verification and billing requests?

**Comment by SebastianRehfeldt** [comment]
> @SebastianRehfeldt @fongie Thanks for making this documentation + library, its a massive help! This is really well structured information on both the verification and billing process.

Hey @mkhq ! Happy that it's helping :)

> One thing that was a bit unclear to me are the options for when requests are failing, e.g. code 201, 202, connection drops, no response, ... . Can we retry both the verification and billing requests?

Good question. I just got started this week with making test requests to the diga api and did not experience any 201/202 error responses. I'd expect that we can retry the requests but have not read anything about it in the official documentation or tested myself.
As the document from this PR contains an FAQ section I can add your question to it and maybe someone else will have an answer for it.

**Comment by mkhq** [comment]
> Good question. I just got started this week with making test requests to the diga api and did not experience any 201/202 error responses. I'd expect that we can retry the requests but have not read anything about it in the official documentation or tested myself.

Yes, I also dug into the official docs and couldn't find anything. Only the note about expected availability in sec. 5.6, 98% between 8-20 Mon-Fri and 90% for the rest. With those SLOs I'd expect a couple of failed requests now and then.

**Comment by SebastianRehfeldt** [comment]
> Yes, I also dug into the official docs and couldn't find anything. Only the note about expected availability in sec. 5.6, 98% between 8-20 Mon-Fri and 90% for the rest. With those SLOs I'd expect a couple of failed requests now and then.

We also stumbled across the response/availability times. But from the response times from our test requests and feedback from @fongie I think the requests respond rather quick (definitely <1 sec). Should hopefully not be an issue 🤞

**Comment by fongie** [comment]
Hi @mkhq that is a good question. I think you can safely build an implementation that retries failed requests at least a few times. Maybe @gtuk can shed some light how they do it since he's running in production? 😃

**Comment by gtuk** [comment]
Sorry for the late reply was kinda busy the last days. Here are some infos from running it in production for a while no:
- Response times vary a lot between the different insurance endpoints (1-8 sec) (the ones from the bitmark are the fastest)
- Response times between the validation and the billing endpoint are also significant for most of the insurance endpoints (billing takes around 1-4 sec more)
- Retrying the validation and the billing multiple times is in principal no problem as the insurances handle these cases for you
- Not all insurance implementations/endpoints are following the spec completely (I could only verify that the bitmark endpoints are following it, e.g dateOfProvision is only checked by them for correctness)
- We are only retrying the billing process automatically not the validation (it is a custom logic with an exponential backoff)

I hope this helps already a bit

**Comment by SebastianRehfeldt** [comment]
Thanks @gtuk for the insights from running into production! Just added a quick section about response times to the FAQ section of the documentation

**Comment by SebastianRehfeldt** [comment]
Hi @fongie, Any chance we can get this PR merged/closed soon? :)

**Comment by fongie** [comment]
hi @SebastianRehfeldt just got back from vacation, will look into this soon

**Review by fongie — APPROVED** [review]
thank you for all your work in gathering and presenting this information! it looks great, and please let's try to keep it up to date as we go along or they change things up

**Comment by SebastianRehfeldt** [comment]
> let's try to keep it up to date as we go along or they change things up

Agree, let's do! :)

**Pull request merged**
